src/server/game/Entities/Player/Player.cpp

```diff
@@ -26101,6 +26108,160 @@ void Player::ResetTalentSpecialization()
+}
+
+bool Player::AddPvpTalent(PvpTalentEntry const* talent, uint8 activeTalentGroup, bool learning)
+{
+    ASSERT(talent);
+    SpellInfo const* spellInfo = sSpellMgr->GetSpellInfo(talent->SpellID);
+    if (!spellInfo)
+    {
+        TC_LOG_ERROR("spells", "Player::AddTalent: Spell (ID: %u) does not exist.", talent->SpellID);
+        return false;
+    }
+
+    if (!SpellMgr::IsSpellValid(spellInfo, this, false))
+    {
+        TC_LOG_ERROR("spells", "Player::AddTalent: Spell (ID: %u) is invalid", talent->SpellID);
+        return false;
+    }
+
+    if (IsInAreaThatActivatesPvpTalents())
+        LearnSpell(talent->SpellID, false);
+
+    // Move this to toggle ?
+    if (talent->OverridesSpellID)
+        AddOverrideSpell(talent->OverridesSpellID, talent->SpellID);
+
+    PlayerTalentMap::iterator itr = GetPvpTalentMap(activeTalentGroup)->find(talent->ID);
+    if (itr != GetPvpTalentMap(activeTalentGroup)->end())
+        itr->second = PLAYERSPELL_UNCHANGED;
+    else
+        (*GetPvpTalentMap(activeTalentGroup))[talent->ID] = learning ? PLAYERSPELL_NEW : PLAYERSPELL_UNCHANGED;
+
+    return true;
+}
+
+void Player::RemovePvpTalent(PvpTalentEntry const* talent)
+{
+    SpellInfo const* spellInfo = sSpellMgr->GetSpellInfo(talent->SpellID);
+    if (!spellInfo)
+        return;
+
+    RemoveSpell(talent->SpellID, true);
+
+    // Move this to toggle ?
+    if (talent->OverridesSpellID)
+        RemoveOverrideSpell(talent->OverridesSpellID, talent->SpellID);
+
+    // if this talent rank can be found in the PlayerTalentMap, mark the talent as removed so it gets deleted
+    PlayerTalentMap::iterator plrPvpTalent = GetPvpTalentMap(GetActiveTalentGroup())->find(talent->ID);
+    if (plrPvpTalent != GetPvpTalentMap(GetActiveTalentGroup())->end())
+        plrPvpTalent->second = PLAYERSPELL_REMOVED;
+}
+
+void Player::TogglePvpTalents(bool enable)
+{
+    PlayerTalentMap const* pvpTalents = GetPvpTalentMap(GetActiveTalentGroup());
+    for (PlayerTalentMap::value_type const& v : *pvpTalents)
+    {
+        PvpTalentEntry const* pvpTalentInfo = sPvpTalentStore.AssertEntry(v.first);
+        if (enable && v.second != PLAYERSPELL_REMOVED)
+            LearnSpell(pvpTalentInfo->SpellID, false);
+        else
+            RemoveSpell(pvpTalentInfo->SpellID, true);
+    }
+}
+
+bool Player::HasPvpTalent(uint32 talentID, uint8 activeTalentGroup) const
+{
+    PlayerTalentMap::const_iterator itr = GetPvpTalentMap(activeTalentGroup)->find(talentID);
+    return (itr != GetPvpTalentMap(activeTalentGroup)->end() && itr->second != PLAYERSPELL_REMOVED);
+}
+
+bool Player::IsInAreaThatActivatesPvpTalents() const
+{
+    if (InBattleground())
+        return true;
+
+    uint32 zoneID, areaID;
+    GetZoneAndAreaId(zoneID, areaID);
+
+    return IsAreaThatActivatesPvpTalents(areaID);
+}
+
+bool Player::IsAreaThatActivatesPvpTalents(uint32 areaID) const
+{
+    if (AreaTableEntry const* area = sAreaTableStore.LookupEntry(areaID))
```

Comment:
I got something (local) for the AuraScript and using the spell_area but I noticed that when we cast the spell outside the areas specified in spell_area table we get SPELL_FAILED_INCORRECT_AREA. So I was thinking about just having a function called EnablesPvpTalents in AreaTableEntry instead of IsPvpZone which will check for the zoneID's... it's not perfect tho and will require some more checks for wintergrasp & tol barad (although maybe we can let scripts handle that).